releasing-archway.md

```diff
@@ -1,74 +1,41 @@
-    the release-X.Y branch, as needed.
-  - X.Y.Z is cut straight from the release-X.Y branch, and X.Y.Z+1-beta.0 is
-    tagged on the followup commit.
-- X.Y.Z, Z > 0 (Branch: release-X.Y.Z) (Branched [patch releases](#patch-releases) only for hotfix situations)
-  - These are rarely used and are special in that the X.Y.Z tag is branched to isolate
-    the emergency/critical fix from all other changes that have landed on the
-    release branch since the previous tag
-  - Cut release-X.Y.Z branch to hold the isolated patch release
-  - Tag release-X.Y.Z branch + fixes with X.Y.(Z+1)
-  - Branched [patch releases](#patch-releases) are rarely needed but used for
-    emergency/critical fixes to the latest release
-
-### Major version timeline
-
-There is currently no mandated timeline for major versions beyond version 1.Y.Z, Y,Z >= 0. We haven't so far applied a rigorous interpretation of semantic
-versioning with respect to incompatible changes of any kind.
-
-TODO: Major versioning criteria need to be put up for discussion once Archway protocol reaches 1.Y.Z release.
-
-## Patch releases
-
-Patch releases are intended for critical bug fixes to the latest minor version,
-such as addressing security vulnerabilities, fixes to problems affecting a large
-number of users and severe problems with no workaround.
-
-They should not contain miscellaneous feature additions or improvements, and
-especially no incompatibilities should be introduced between patch versions of
-the same minor version (or even major version).
-
-Dependencies, such as cosmos-sdk or tendermint, should also not be changed unless
-absolutely necessary, and also just to fix critical bugs (so, at most patch
-version changes, not new major nor minor versions).
+# Archway Protocol Release Process
+
+## Release Versioning
+
+This document outlines the release process for Archway protocol software.
+
+The **Archway** protocol follows a **state versioning system**, it does not utilize traditional [semantic versioning](http://semver.org); SemVer is specifically about APIs, but API breaking changes is not the only way in which libraries can "break". In deterministic systems, such as a blockchain being deterministic state machine, *state breaking changes far outweigh API breaking changes*. As such, Archway follows a State Versioning system.
+
+Given a version number Major.Minor.Patch, increment the:
+
+1. **Major** version when any state breaking changes are introduced;
+2. **Minor** version when any API changes, both API-compatible or API-incompatible, are introduced;
+3. **Patch** version when any state-compatible bug fixes are introduced;
+
+Additional labels for release-candidates, pre-release versions and other build metadata are available as extensions to the Major.Minor.Patch format.
+
+**Note:** Any dependency updates that are not state breaking, e.g. updating the Go version, fall under minor.
```

Comment:
If yes, would similar logic apply for consensus breaking changes? Do we tag a major version as soon as a state breaking PR is merged? (Tangent: Would be nice to have some gh workflow which checks that when there is a major version bump, an upgrade handler for that version must exist)

If no, does that mean we should batch the non state breaking changes into single release? What would be the release schedule in such case?

Reply:
These are great questions! It would largely depend on the perceived impact of the update across the ecosystem. 

Consider for example if we do a non-state breaking, but API breaking change; We probably don't want to immediately tag and release this to provide developers time to integrate and test. Thus, we might prefer to do such a release more akin to a major update with some formal coordination;

Conversely, if we update the Go version, which would result in a minor version bump, we could immediately tag such a release as its impact would be mostly superficial;

I do like the idea of having a workflow to check for an upgrade handler for that version 👍🏻 

In general, I believe we should be more selective with major version releases and handle them on a case by case basis. IMHO blockchain release is more akin to traditional software pre-Internet era, as patches and fixes could not easily be rolled out. Thus greater care needs to be taken with each major release as the "move fast and break things" approach is disconcerting in an immutable state (blockchain) environment.

Reply:
Regarding the distinction between binary updates vs non-binary updates...

We could simply add any non-binary updates, e.g. docs, workflows, etc. under unreleased version until we have an actual binary update and tag along with the updated binary release; I believe this will provide for a more consistent upgrade path where every release is associated with a binary update.